### PR TITLE
Initial commit of primitive_array type

### PR DESCRIPTION
**Pull request by liamembark**
Fixes: DATA-556

Description of Changes
======================
Using pre-allocated `RosValue`s to hold each primitive within an array can be very expensive for large arrays of small data types.
This change makes it so that we create a RosValue on the stack to point to an element of these arrays whenever the items are accessed. This involves a few steps:

1. `RosValue::Pointer`s now have two 'underlying definitions': one where we track the location in the dynamic vector, and one where we simply store a `RosValue`
2. Added a `primitive_array` type which stores primitives (not including strings) which do not have `RosValue`s pre-allocated for them.

Test Plan
=========
Verified that a large amount of memory was no longer being used by reading a bag full of CompressedImage messages.

### CODE REVIEW

File: lib/ros_value.h [hunk 1]
```diff
@@ -251,24 +239,27 @@ class RosValue {
...
     } else if (type_ == Type::array) {
       array_info_.~array_info_t();
+    } else if (type_ == Type::primitive_array) {
+      primitive_array_info_.~primitive_array_info_t();
     } else {
       primitive_info_.~primitive_info_t();
     }
   }
+  RosValue operator=(const RosValue&) {
+    throw std::runtime_error("Can't call this");
```
Comment (brendangeck) [inline lib/ros_value.h:249]: I feel like we should have a better error message than this

**Review by brendangeck — CHANGES_REQUESTED** [review]
General feedback: This is now getting to the point where it's difficult for me to "simulate" the codepath in my head as I review the PR. I think in cases like these, I'd prefer we write some tests to confirm this logic will continue to work after future changes.

File: lib/ros_value.h (hunk 1 above)
Comment (liamembark) [reply lib/ros_value.h:249]: I actually want to write this, thanks for catching.

**Comment by liamembark** [comment]
> General feedback: This is now getting to the point where it's difficult for me to "simulate" the codepath in my head as I review the PR. I think in cases like these, I'd prefer we write some tests to confirm this logic will continue to work after future changes.

@brendangeck I can definitely add some more tests around arrays specifically. The existing tests do a decent job of covering the `primitive_array` type, but not so much normal arrays and some other edge cases.

**Review by brendangeck — APPROVED** [review]
